File: Exercicios-JAVA/Filas/FilaFlexivel/FilaFlexivel.java
import java.util.Scanner;

public class FilaFlexivel {
    class Celula{
        private int valor;
        private Celula proximo;

        public Celula(int valor){
            this.valor = valor;
            this.proximo = null;
        }

        public int getValor(){
            return valor;
        }
    }

    private Celula inicio;
    private Celula fim;

    public FilaFlexivel(){
        this.inicio = null;
        this.fim = null;
    }

    public void enqueue(int valor){

        Celula celula = new Celula(valor);

        if(inicio == null){

            inicio = celula;

        }else{

            fim.proximo = celula;
        }

        fim = celula;
    }

    public int dequeue(){
        if(this.inicio == null){
            System.out.println("Fila Vazia");
        }

        int valor = this.inicio.getValor();
        this.inicio = this.inicio.proximo;

        if(this.inicio == null){
            this.fim = null;
        }


        return valor;

    }

    public void mostrar(){
        if(inicio == null){
            System.out.println("Fila Vazia");
        }

        Celula i = inicio;

        System.out.print("Fila: ");
        while(i != null){
            System.out.print(i.getValor() + " ");
            i = i.proximo;
        }
        System.out.println();
    }


    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        FilaFlexivel fila = new FilaFlexivel();
        int opc;

        do { 
            System.out.println("----- MENU -----");
            System.out.println("Digite 0 para SAIR");
            System.out.println("Digite 1 para INSERIR");
            System.out.println("Digite 2 para DELETAR");
            System.out.println("Digite 3 para MOSTRAR");
            System.out.print("Escolha uma opcao: ");

            opc = scanner.nextInt();

            switch(opc) {
                case 1:{
                    System.out.print("Digite o valor: ");
                    int valor = scanner.nextInt();
                    fila.enqueue(valor);
                    break;
                }
                case 2:{
                    int resul = fila.dequeue();
                    System.out.println("Valor removido: " + resul);
                    break;
                }
                case 3:
                    fila.mostrar();
                    break;
                default:
                    throw new AssertionError();
            }
            
        } while (opc != 0);

    }

}

File: Exercicios C/Filas/FilaFlexivel/FilaFlexivel.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Celula{
    int valor;
    struct Celula *proximo;
}Celula;

typedef struct FilaFlexivel{

    Celula *inicio;
    Celula *fim;

}FilaFlexivel;

FilaFlexivel *criarFila(){ // "Construtor" celula cabeca
    FilaFlexivel *fila = (FilaFlexivel*)malloc(sizeof(FilaFlexivel));
    Celula *cabeca = (Celula*)malloc(sizeof(Celula));
    cabeca->valor = 0;
    cabeca->proximo = NULL;

    fila->inicio= cabeca;
    fila->fim = fila->inicio;

    return fila;
    
}

void enqueue(FilaFlexivel *fila, int x){
    Celula *celula = (Celula*)malloc(sizeof(Celula));
    celula->valor = x;
    celula->proximo = NULL;

    fila->fim->proximo = celula;
    fila->fim = celula;
    
}

int dequeue(FilaFlexivel *fila){
    if(fila->inicio == fila->fim){
        printf("Fila vazia!");
    }
    Celula *tmp = fila->inicio->proximo;
    int valorRemover = tmp->valor;
    fila->inicio->proximo = tmp->proximo;

    free(tmp);
    
    return valorRemover;
}

void mostrar(FilaFlexivel *fila){
    if(fila->inicio == fila->fim){
        printf("Fila vazia!");
    }

    Celula *i = fila->inicio->proximo;
    printf("Fila: ");
    while(i != NULL){
        printf("%d ", i->valor);
        i = i->proximo;
    }
    printf("\n");
    
}

int main(){

    FilaFlexivel *fila = criarFila();
    int opc;

    do { 
            printf("\n----- MENU -----");
            printf("\nDigite 0 para SAIR");
            printf("\nDigite 1 para INSERIR");
            printf("\nDigite 2 para DELETAR");
            printf("\nDigite 3 para MOSTRAR");
            printf("\nEscolha uma opcao: ");

            scanf("%d", &opc);

            switch(opc) {
                case 1:{
                    printf("\nDigite o valor: ");
                    int valor; 
                    scanf("%d", &valor);
                    enqueue(fila,valor);
                    break;
            }
                case 2:{
                    int resul = dequeue(fila);
                    printf("\nValor removido: %d", resul);
                    break;
            }
                case 3:
                    mostrar(fila);
                    break;
        }
            
    } while (opc != 0);

    return 0;
}